src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        if (action.extraData && typeof action.extraData === 'object') {
+          currentState = { ...currentState, ...action.extraData };
+        } else {
+          throw new Error('Invalid extraData in addProperties action');
+        }
+        break;
+
+      case 'removeProperties':
+        if (Array.isArray(action.keysToRemove)) {
+          currentState = { ...currentState };
```

Comment:
The line `currentState = { ...currentState };` is redundant here because `currentState` is already a clone of itself. You can remove this line to simplify the code.